std/range/package.d

```diff
@@ -2553,6 +2557,121 @@ auto takeNone(R)(R range)
+ + $(WEB en.wikipedia.org/wiki/Tail_%28Unix%29, _tail) utility. When the length
+ + of $(D _range) is less than or equal to $(D n), $(D _range) is returned
+ + as-is.
+ +
+ + Completes in $(BIGOH 1) steps for ranges that support slicing and have
+ + length. Completes in $(BIGOH _range.length) time for all other ranges.
+ +/
+Range tail(Range)(Range range, size_t n)
+    if (isForwardRange!Range && !isInfinite!Range)
+{
+    static if (hasLength!Range)
+    {
+        // Evaluates `length` fewer times than calling `drop` would in the
+        // slicing case
+        immutable length = range.length;
+
+        if (n >= length)
+            return range;
+
+        static if (hasSlicing!Range)
+            return range[length - n .. length];
+        else
+            return .drop(range, length - n);
+    }
+    else
+    {
+        if (range.empty)
+            return range;
+
+        Range scout = range.save;
+        for (size_t i = 0; i < n; ++i)
+        {
+            scout.popFront();
+            if (scout.empty)
+                return range;
+        }
+
+        auto tailRange = range.save;
+        do
+        {
+            scout.popFront();
+            tailRange.popFront();
+            assert(!tailRange.empty);
+        }
+        while(!scout.empty);
+
+        return tailRange;
+    }
+}
+
+/// Ditto
+Range tail(Range)(Range range, size_t n)
+    if (isInputRange!Range && !isForwardRange!Range && hasLength!Range)
+{
+    return .drop(range, range.length - n);
+}
+
+
+///
+pure @safe unittest
+{
+    // tail -c n
+    assert([1, 2, 3].tail(1) == [3]);
+    assert([1, 2, 3].tail(2) == [2, 3]);
+    assert([1, 2, 3].tail(3) == [1, 2, 3]);
+    assert([1, 2, 3].tail(4) == [1, 2, 3]);
+    assert([1, 2, 3].tail(0).length == 0);
+
+    // tail --lines=n
+    import std.algorithm.comparison : equal;
+    import std.algorithm.iteration : joiner;
+    import std.string : lineSplitter;
+
+    assert("one\ntwo\nthree"
+            .lineSplitter
+            .tail(2)
+            .joiner("\n")
+            .equal("two\nthree"));
+}
+
+// @nogc prevented by @@@BUG@@@ 15408
+pure nothrow @safe /+@nogc+/ unittest
+{
+    import std.algorithm.comparison : equal;
+    import std.internal.test.dummyrange;
+
+    static immutable cheatsheet = [6, 7, 8, 9, 10];
+
+    foreach (R; AllDummyRanges)
+    {
+        static if ((isForwardRange!R && !isInfinite!R) || // First overload
+                   (isInputRange!R && hasLength!R)) // Second overload
+            assert(R.init.tail(5).equal(cheatsheet));
+    }
+
+    // Infinite ranges are not supported
+    static assert(!__traits(compiles, repeat(0).tail(0)));
+
+    // Neither are non-forward ranges without length
+    static assert(!__traits(compiles, DummyRange!(ReturnBy.Value, Length.No,
+        RangeType.Input).init.tail(5)));
+}
+
+@nogc unittest
+{
+    static immutable input = [1, 2, 3];
+    static immutable expectedOutput = [2, 3];
+    assert(input.tail(2) == expectedOutput);
```

Comment:
If you change it to a static assert, you can test CTFE and @nogc at the same time, right?